news/_posts/2020-11-30-from-legacy-to-future-architecture-connecting-the-dots.md

```diff
@@ -0,0 +1,308 @@
+
+### An extensible API
+
+PrestaShop is a development platform, and as such it requires a balance between extensibility and predictability. In an ecosystem as rich as PrestaShop's, it is impossible to foresee the needs of each individual project in advance; the Core API must then be **open for extension** so that new features can be added as needed. At the same time, extension needs the system to behave in a predictable way; in consequence the API must also be **closed to modification**. As you might have guessed, I'm talking about the O in SOLID: the [Open/Closed principle][open-closed-principle].
+
+This principle essentially states that objects should be made in such a way that you can add new functionality without modifying the existing code. In the engineering literature, Bertrand Meyers originally suggested achieving this through subclassing, while Robert C. Martin encourages doing it through interface reimplementation. The Core API leverages CQRS and Command & Query buses in the spirit of this.
+
+The API's "interface" is composed of **Commands** that describe an action that can be performed by the system, **Queries** that define a data retrieval operation, and **Query Result Types** (also commonly described as Data Transfer Objects or DTOs), which specify the structure for the data returned by a given Query operation. Since these objects (in particular Commands and Queries) are hardwired throughout the project, they can be considered closed to modification.
+
+The _implementation_, on the other hand, is performed by Handler services which _subscribe_ to a given Command or Query. These services can be easily rewired through [symfony service decoration][sf-service-decoration], so any extension can redefine the behavior of any given use case by decorating the original handler or reimplementing its interface. Additionally, in some cases, multiple extensions can decorate the same service through a _decoration chain_, where each services "hooks" into the input or output of the previous decorator.
+
+The Core API is the better replacement for the obsolete [legacy class override system][class-override], which will progressively be phased out as legacy classes are removed from the root namespace.
+
+### A domain-oriented API
+
+Consider the following application use cases:
+
+- Create a customer through a form in BO.
+- Load customer information for update.
+- Issue a partial refund in an order.
+- Activate debug mode.
+
+Each one of these is a unique and complete user interaction with the system. Whenever a request is issued to the backend, it can usually be matched with exactly one use case. 
+
+The Core API is built around these use cases. This means that all its operations are  based on the things the software does from a user standpoint, rather than what its _underlying components_ are able to do. This is a subtle, yet very important difference.
```

Comment:
```suggestion
The Core API is built around these use cases. This means that all its operations are based on the things the software does from a user standpoint, rather than what its _underlying components_ are able to do. This is a subtle, yet very important difference.
```